std/range/package.d

```diff
@@ -10761,3 +10764,477 @@ pure unittest
+        {
+            static if (hasSwappableElements!Range)
+                auto ref opIndex(size_t index) inout
+                {
+                    if (hasFallback) return fallback;
+                    return r[index];
+                }
+            else
+            {
+                auto opIndex(size_t index) inout
+                {
+                    if (hasFallback) return fallback;
+                    return r[index];
+                }
+            }
+        }
+
+        static if (isRandomAccessRange!Range && hasAssignableElements!Range)
+        void opIndexAssign()(auto ref ElementType!Range val, size_t n)
+        {
+            if (hasFallback)
+            {
+                assert(n <= 1, "fallback contains only one element.");
+                fallback = val;
+            }
+            else
+            {
+                r[n] = val;
+            }
+        }
+
+        static if (hasLength!Range)
+        {
+            @property size_t length() const
+            {
+                if (hasFallback) return 1;
+                return r.length;
+            }
+
+            alias opDollar = length;
+        }
+
+        static if (hasSlicing!Range)
+        {
+            auto opSlice(size_t i, size_t j)
+            in
+            {
+                assert(
+                    i <= j,
+                    "Attempting to slice a fallback with a larger first argument than the second."
+                );
+            }
+            body
+            {
+                static if (isInfinite!Range)
+                {
+                    // as an infinite type can't be null, this returns opSlice directly
+                    return r[i .. j];
+                }
+                else
+                {
+                    if (hasFallback)
+                    {
+                        assert(j <= 1, "fallback contains only one element.");
+                        auto ret = typeof(this)(r, fallback);
+                        if (i == 1)
+                            ret.hasFallback = false;
+
+                        return ret;
+                    }
+                    else
+                    {
+                        return typeof(this)(r[i .. j], fallback);
+                    }
+                }
+            }
+
+            static if (isInfinite!Range)
+            {
+                private static struct DollarToken {}
+                enum opDollar = DollarToken.init;
+
+                typeof(this) opSlice(size_t i, DollarToken)
+                {
+                    return typeof(this)(r[i .. $], fallback);
+                }
+            }
+        }
+
+        auto opSlice() { return this; }
+
+        string toString()
+        {
+            import std.format : format;
+            if (hasFallback)
+                return "OrElse(%s)".format(fallback);
+            else
+                return "OrElse(%s)".format(r);
+        }
+    }
+
+    return OrElseRange(r, fallback);
```

Comment:
I suggest a dramatically simpler and smaller implementation. Use `r.chain(fallback)` as the range type, and pair it with a `bool` that tells you whether you need to output the last element.

Sadly we don't have a primitive akin to Unix `head` (http://www.computerhope.com/unix/uhead.htm), That utility, if passed a negative number, outputs lines except a last few. If we did have that, the implementation would be a one-liner:

```d
return r.chain(only(fallback)).head(r.empty ? -1 : ssize_t.max);
```

Anyhow, using `chain` should simplify matters quite a bit.